### PR TITLE
[stdlib] Introduce `Byte`

### PR DESCRIPTION
**Pull request by soraros**
And use it in `memory.mojo`. Related to #3634.

### CODE REVIEW

**Review by JoeLoser — APPROVED** [review]

**Comment by JoeLoser** [comment]
LGTM once tests are passing

File: stdlib/test/memory/test_memory.mojo [hunk 1]
```diff
@@ -125,16 +125,16 @@ def test_memcmp():
...
     p2.store(120)
 
-    var c = memcmp(p1, p2, 1)
-    assert_equal(c, -1, "-120 is smaller than 120")
+    c = memcmp(p1, p2, 1)
+    assert_equal(c, 1)
 
     c = memcmp(p2, p1, 1)
-    assert_equal(c, 1, "120 is bigger than -120")
+    assert_equal(c, -1)
```
Comment (soraros) [inline stdlib/test/memory/test_memory.mojo:137]: @JoeLoser I think this small change actually revealed a bug. `memcpy` should treat the underlying buffer as `unsigned char`. [ref](https://cplusplus.com/reference/cstring/memcmp/).

Comment (JoeLoser) [reply stdlib/test/memory/test_memory.mojo:137]: Yeah, I think we discovered this on a PR a few months ago that `memcpy` was always working with signed int8 which was recipe for error.

**Comment by JoeLoser** [comment]
!sync

File: stdlib/test/memory/test_memory.mojo (hunk 1 above)
Comment (soraros) [reply stdlib/test/memory/test_memory.mojo:137]: Sorry I made a typo. `memcpy` using `Int8` is fine (at least from the ABI point of view). It's actually that `memcmp` was also using `Int8`, which is a logical error.

File: stdlib/src/memory/memory.mojo [hunk 2]
```diff
@@ -50,9 +50,10 @@ fn _align_down(value: Int, alignment: Int) -> Int:
 
 
 @always_inline
-fn _memcmp_impl_unconstrained[
+fn _memcmp_impl[
     type: DType
 ](s1: UnsafePointer[Scalar[type], _], s2: __type_of(s1), count: Int) -> Int:
+    constrained[type.is_unsigned(), "the input dtype must be unsigned"]()
```
Comment (soraros) [inline stdlib/src/memory/memory.mojo:56]: Since we use `SIMD.__gt__` etc. in this function, if we don't restrict to `unsigned`, the result will be wrong.

**Comment by soraros** [comment]
@JoeLoser This is a minor change, but I think we should add a changelog entry for it so people are aware and can start using it.

**Comment by JoeLoser** [comment]
> @JoeLoser This is a minor change, but I think we should add a changelog entry for it so people are aware and can start using it.

Agreed, feel free to add a changelog entry to promote its use! 👍

**Comment by soraros** [comment]
@JoeLoser I asked you since this PR is already merged. Is it possible that you add it internally?

File: stdlib/src/memory/memory.mojo [hunk 3]
```diff
@@ -128,14 +121,14 @@ fn memcmp[
...
     @parameter
-    if sizeof[type]() >= sizeof[DType.int32]():
+    if sizeof[type]() >= sizeof[UInt32]():
         return _memcmp_impl(
-            s1.bitcast[Int32](),
-            s2.bitcast[Int32](),
-            byte_count // sizeof[DType.int32](),
+            s1.bitcast[UInt32](),
+            s2.bitcast[UInt32](),
+            byte_count // sizeof[UInt32](),
```
Comment (soraros) [inline stdlib/src/memory/memory.mojo:128]: This is very unsafe and wrong.

**Comment by JoeLoser** [comment]
> @JoeLoser I asked you since this PR is already merged. Is it possible that you add it internally?

Oh I see, yeah, I just opened a PR internally for a changelog entry - no biggie 👍

**Pull request closed without merging**